jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,514 @@
+
+    return modified;
+  }
+};
+
+using ConstraintVariant = std::variant<AllPointeesPointToSupersetConstraint, SupersetOfAllPointeesConstraint>;
+
+/**
+ * A class for adding and applying constraints to the points-to-sets of the PointerObjectSet
+ * Some constraints are applied immediately, while others are kept in lists for later constraint solving.
+ * Use Solve() to calculate the final points-to-sets.
+ *
+ * Constraints on the special external and escaped nodes are built in.
+ */
+class PointerObjectConstraintSet final
+{
+public:
+  explicit PointerObjectConstraintSet(PointerObjectSet& set) : Set_(set) {}
+
+  PointerObjectConstraintSet(const PointerObjectConstraintSet& other) = delete;
+
+  PointerObjectConstraintSet&
+  operator =(const PointerObjectConstraintSet& other) = delete;
+
+  /**
+   * The simplest constraint, on the form: pointee in P(pointer)
+   */
+  void
+  AddPointerPointeeConstraint(PointerObjectIndex pointer, PointerObjectIndex pointee)
+  {
+    JLM_ASSERT(pointer <= Set_.NodeCount());
+    JLM_ASSERT(pointee <= Set_.NodeCount());
+
+    // All set constraints are additive, so simple constraints like this can be directly applied and forgotten.
+    Set_.AddToPointsToSet(pointer, pointee);
+  }
+
+  void
+  AddConstraint(ConstraintVariant c)
+  {
+    Constraints_.push_back(c);
+  }
+
+  /**
+   * Iterates over and applies constraints until the all points-to-sets satisfy them
+   */
+  void
+  Solve();
+
+private:
+
+  /**
+   * Internal function used to implement the constraints on flags
+   */
+  bool
+  ApplyEscapedFlagPropagation();
+
+  // The PointerObjectSet being built upon
+  PointerObjectSet& Set_;
+
+  // Lists of all constraints, of all different types
+  std::vector<ConstraintVariant> Constraints_;
+};
+
+void
+PointerObjectConstraintSet::Solve()
+{
+  // Keep applying constraints until no sets are modified
+  bool modified = true;
+  while (modified) {
+    modified = false;
+
+    for (auto& constraint : Constraints_)
+      std::visit([&](auto constraint) {
+        modified |= constraint.Apply(Set_);
+      }, constraint);
+
+    modified |= ApplyEscapedFlagPropagation();
+  }
+}
+
+bool
+PointerObjectConstraintSet::ApplyEscapedFlagPropagation()
+{
+  // This function only applies one iteration of escape flag propagation, so it might need to be run again
+  bool modified = false;
+
+  for (PointerObjectIndex idx = 0; idx < Set_.NodeCount(); idx++)
+  {
+    if (!Set_.GetNode(idx).HasEscaped())
+      continue;
+
+    for (PointerObjectIndex pointee : Set_.GetPointsToSet(idx)) {
+      auto& pointeeNode = Set_.GetNode(pointee);
+      modified |= !pointeeNode.HasEscaped();
+      pointeeNode.MarkAsEscaped();
+    }
+  }
+
+  return modified;
+}
+
+std::unique_ptr<PointsToGraph>
+Andersen::Analyze(const RvsdgModule &module, jlm::util::StatisticsCollector &statisticsCollector)
+{
+  Set_ = std::make_unique<PointerObjectSet>();
+  Constraints_ = std::make_unique<PointerObjectConstraintSet>(*Set_);
```

Comment:
Please call `reset()` on the `Set_` and `Constraints_` unique pointers before you exit the function.